From ea90063cbcc03506510522809d564112aa1a9ce9 Mon Sep 17 00:00:00 2001
From: Willy Tarreau <w@1wt.eu>
Date: Tue, 17 Dec 2013 14:21:48 +0100
Subject: [PATCH 2/2] BUG/MEDIUM: stream-int: fix the keep-alive idle
 connection handler

Commit 2737562 (MEDIUM: stream-int: implement a very simplistic idle
connection manager) implemented an idle connection handler. In the
case where all data is drained from the server, it fails to disable
polling, resulting in a busy spinning loop.

Thanks to Sander Klein and Guillaume Castagnino for reporting this bug.

No backport is needed.
Index: haproxy-1.5-dev21/src/stream_interface.c
===================================================================
--- haproxy-1.5-dev21.orig/src/stream_interface.c
+++ haproxy-1.5-dev21/src/stream_interface.c
@@ -508,6 +508,8 @@ static void si_idle_conn_null_cb(struct 
 	/* disable draining if we were called and have no drain function */
 	if (!conn->ctrl->drain)
 		__conn_data_stop_recv(conn);
+	else if (!(conn->flags & CO_FL_SOCK_RD_SH))
+		__conn_data_poll_recv(conn);
 }
 
 /* Callback to be used by connection I/O handlers when some activity is detected
